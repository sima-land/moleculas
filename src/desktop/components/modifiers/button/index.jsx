import React, { forwardRef } from 'react';
import isNumber from 'lodash/isNumber';
import classes from './modifier-button.scss';
import classnames from 'classnames/bind';
import { MODIFIERS_TYPES as TYPES } from '@dev-dep/ui-nucleons/constants';

const cx = classnames.bind(classes);

const typesList = Object.values(TYPES);

/**
 * Возвращает компонент модификатора.
 * @param {Object} props Свойства компонента.
 * @param {number|null} [props.count=null] Количество компонента.
 * @param {string} props.content Содержимое модификатора.
 * @param {string} [props.image] URL изображения модификатора.
 * @param {'text'|'image'} [props.type='text'] Тип содержимого модификатора.
 * @param {boolean} [props.selected] Выбран ли модификатор.
 * @param {string} props.color Цвет модификатора.
 * @param {string} [props.className] Дополнительный CSS-класс.
 * @param {Function} [props.onClick] Сработает при клике на модификатор.
 * @param {boolean} [props.isMarkdown] Имеет ли товар уценку.
 * @param {boolean} [props.squared] Должна ли быть кнопка квадратной.
 * @param {Object} ref Реф.
 * @return {ReactElement} Компонент модификатора.
 */
export const ModifierButton = forwardRef(function ModifierButton ({
  count = null,
  content,
  selected,
  color,
  image,
  type: typeProp = 'text',
  className,
  onClick,
  isMarkdown: hasMarkdown,
  squared = [TYPES.color, TYPES.image].includes(typeProp),
}, ref) {
  const type = typesList.includes(typeProp) ? typeProp : 'text';
  const hasCount = isNumber(count) && count >= 0;

  return (
    <div
      ref={ref}
      className={cx(
        'modifier-button',
        squared && 'squared',
        selected && 'selected',
        className
      )}
      style={color ? { background: color } : undefined}
      onClick={onClick}
    >
      {/* label */}
      {(hasCount || hasMarkdown) && (
        <span
          className={cx(
            'label',
            hasCount && 'with-count'
          )}
        >
          {hasCount && (count > 99 ? '99+' : count)}
          {hasMarkdown && (hasCount ? '(У)' : 'У')}
        </span>
      )}

      {/* content */}
      {[TYPES.text, TYPES.image].includes(type) && (
        <div className={cx('content')}>
          {type === TYPES.text && (
            <span className={cx('text', !squared && 'guttered')}>
              {String(content || '')}
            </span>
          )}

          {type === TYPES.image && (
            <img
              src={image}
              alt={content}
              className={cx('image')}
            />
          )}
        </div>
      )}
    </div>
  );
});
