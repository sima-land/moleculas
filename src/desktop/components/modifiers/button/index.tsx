import { CSSProperties, MouseEventHandler, forwardRef } from 'react';
import { isNumber } from 'lodash';
import classes from './modifier-button.module.scss';
import classnames from 'classnames/bind';
import { ModifierType } from '../../../../common/types';
import { MODIFIER_TYPE as TYPE } from '../../../../common/constants';

export interface Modifier {
  count?: number;
  content: string;
  color?: string;
  image?: string;
  type?: ModifierType;
  isMarkdown?: boolean;
  selected?: boolean;
}

export interface ModifierButtonProps extends Modifier {
  className?: string;
  style?: CSSProperties;
  onClick?: MouseEventHandler<HTMLDivElement>;
  squared?: boolean;
  'data-testid'?: string;
}

const cx = classnames.bind(classes);

/**
 * Компонент модификатора.
 */
export const ModifierButton = forwardRef<HTMLDivElement, ModifierButtonProps>(
  function ModifierButton(
    {
      count = null,
      content,
      selected,
      color,
      image,
      type = 'text',
      className,
      style,
      onClick,
      isMarkdown: hasMarkdown,
      squared = [TYPE.color, TYPE.image].includes(type),
      'data-testid': testId = 'modifier-button',
    },
    ref,
  ) {
    const hasCount = isNumber(count) && count >= 0;

    return (
      <div
        ref={ref}
        className={cx('modifier-button', squared && 'squared', selected && 'selected', className)}
        style={color ? { ...style, background: color } : style}
        onClick={onClick}
        data-testid={testId}
      >
        {/* label */}
        {(hasCount || hasMarkdown) && (
          <span
            className={cx('label', hasCount && 'with-count')}
            data-testid='modifier-button:badge'
          >
            {hasCount && ((count as any) > 99 ? '99+' : count)}
            {hasMarkdown && (hasCount ? '(У)' : 'У')}
          </span>
        )}

        {/* content */}
        {[TYPE.text, TYPE.image].includes(type) && (
          <div className={cx('content')}>
            {type === TYPE.text && (
              <span className={cx('text', !squared && 'guttered')}>{String(content || '')}</span>
            )}

            {type === TYPE.image && <img src={image} alt={content} className={cx('image')} />}
          </div>
        )}
      </div>
    );
  },
);
