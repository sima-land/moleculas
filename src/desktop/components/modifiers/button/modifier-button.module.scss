@use 'node_modules/@sima-land/ui-nucleons/colors';
@import 'node_modules/@sima-land/ui-nucleons/styling/shadows.module';
@import '../modifiers';

.modifier-button {
  max-width: 100%;
  position: relative;
  display: inline-block;
  border-radius: 4px;
  min-width: $minSize;
  height: $minSize;
  cursor: pointer;
  &.squared {
    width: $minSize;
  }
  &::before {
    content: '';
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    border-radius: inherit;
    box-shadow: inset 0 0 0 1px colors.$basic-gray12;
  }
  &:hover::before {
    box-shadow: inset 0 0 0 2px colors.$basic-gray12;
  }
  &.selected::before {
    box-shadow: inset 0 0 0 2px colors.$basic-blue;
  }
}

.content {
  display: flex;
  align-items: center;
  justify-content: center;
  overflow: hidden;
  border-radius: inherit;
  height: 100%;
}

.image {
  display: block;
  width: auto;
  height: calc(100% - 2px);
}

.text {
  text-overflow: ellipsis;
  white-space: nowrap;
  overflow: hidden;
  &.guttered {
    margin: 0 16px;
  }
}

.label {
  position: absolute;
  top: -4px;
  right: -4px;
  display: block;
  min-width: 16px;
  height: 16px;
  text-align: center;
  padding: 0 4px;
  line-height: 16px;
  border-radius: 32px;
  background-color: #fff;
  color: colors.$basic-gray54;
  font-size: 12px;
  font-weight: 500;
  box-sizing: border-box;
  &.with-count {
    background-color: colors.$basic-blue;
    color: #fff;
  }
  &:not(.with-count) {
    @extend .shadow-z1;
  }
}
