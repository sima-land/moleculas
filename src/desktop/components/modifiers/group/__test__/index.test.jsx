import React from 'react';
import { mount, shallow } from 'enzyme';
import {
  ModifiersGroup,
  createHiddenNodeIndexFinder,
} from '..';
import { ModifierButton } from '../../button';
import prop from 'lodash/fp/prop';

jest.mock('..', () => {
  const original = jest.requireActual('..');

  return {
    ...original,
    __esModule: true,
    defaultOnSelectItem: jest.fn(original.defaultOnSelectItem),
    createHiddenNodeIndexFinder: jest.fn(original.createHiddenNodeIndexFinder),
  };
});

describe('<ModifiersGroup />', () => {
  const getTestItems = () => [
    { content: 'Test 1' },
    { content: 'Test 2' },
    { content: 'Test 3' },
    { content: 'Test 4' },
    { content: 'Test 5' },
  ];

  it('should render without props', () => {
    const wrapper = mount(
      <ModifiersGroup />
    );

    expect(wrapper.find('.modifiers-group')).toHaveLength(1);
  });

  it('should call instance.removeGlobalListener on unmount', () => {
    const wrapper = mount(
      <ModifiersGroup />
    );

    const spy = jest.spyOn(wrapper.instance(), 'removeGlobalListener');

    wrapper.unmount();

    expect(spy).toBeCalledTimes(1);
  });

  it('should call checkVisibleNodes() on mount', () => {
    const wrapper = mount(
      <ModifiersGroup />
    );
    jest.spyOn(wrapper.instance(), 'checkVisibleNodes');
    expect(wrapper.instance().checkVisibleNodes).toHaveBeenCalledTimes(0);
    wrapper.instance().componentDidMount();
    expect(wrapper.instance().checkVisibleNodes).toHaveBeenCalledTimes(1);
  });

  it('resetVisibleNodesData() should update state properly', () => {
    const wrapper = mount(
      <ModifiersGroup />
    );
    wrapper.instance().setState({
      hasHiddenNodes: false,
      lastVisibleChildIndex: 10,
    });
    jest.spyOn(document.documentElement, 'offsetWidth', 'get')
      .mockImplementation(() => 500);
    wrapper.instance().resetVisibleNodesData();
    expect(wrapper.instance().state).toMatchObject({
      hasHiddenNodes: true,
      lastVisibleChildIndex: null,
    });
  });

  it('resetVisibleNodesData() shouldn`t update state properly', () => {
    const wrapper = mount(
      <ModifiersGroup />
    );
    wrapper.instance().setState({
      hasHiddenNodes: false,
      lastVisibleChildIndex: 10,
    });
    wrapper.instance().resetVisibleNodesData();
    expect(wrapper.instance().state).toMatchObject({
      hasHiddenNodes: false,
      lastVisibleChildIndex: 10,
    });
  });

  it('checkVisibleNodes() should update state properly', () => {
    const wrapper = mount(
      <ModifiersGroup
        items={getTestItems()}
      />
    );
    const containerEl = wrapper.instance().containerRef.current;
    const showingButtonEl = wrapper.instance().showingButtonRef.current;

    // подменяем значения высоты и отступа сверху у контейнера
    Object.defineProperty(containerEl, 'clientHeight', {
      value: 200,
    });
    Object.defineProperty(containerEl, 'offsetTop', {
      value: 10,
    });

    // подменяем значения высоты и ширины у кнопки показа всех модификаторов
    Object.defineProperty(showingButtonEl, 'clientHeight', {
      value: 50,
    });
    Object.defineProperty(showingButtonEl, 'clientWidth', {
      value: 50,
    });

    // подменяем значения у дочерних элементов (модификаторов) контейнера
    [...containerEl.children].forEach((child, index) => {
      Object.defineProperty(child, 'clientHeight', {
        value: 50,
      });
      Object.defineProperty(child, 'clientWidth', {
        value: 50,
      });

      /*
       * после 5 элемента подменяем у всех элементов offsetTop так,
       * чтобы компоненту ModifiersGroup казалось что он не влез и перенесся
       */
      if (index >= 5) {
        Object.defineProperty(child, 'offsetTop', {
          value: 220,
          configurable: true,
        });
      }
    });

    /*
     * Проверяем ситуацию когда один элемент не влез в высоту контейнера
     * и кнопке НЕ ХВАТИЛО место после последнего влезшего
     */

    // сбрасываем состояние так чтобы метод прошел все условия внутри
    wrapper.instance().resetVisibleNodesData();
    wrapper.update();

    // вызываем метод
    wrapper.instance().checkVisibleNodes();
    wrapper.update();

    expect(wrapper.state()).toMatchObject({
      hasHiddenNodes: true,
      lastVisibleChildIndex: 4,
    });

    /*
     * Проверяем ситуацию когда один элемент не влез в высоту контейнера
     * и кнопке ХВАТИЛО место после последнего влезшего
     */
    jest.spyOn(containerEl, 'getBoundingClientRect').mockImplementation(() => ({ right: 200 }));
    jest.spyOn(containerEl.children[4], 'getBoundingClientRect').mockImplementation(() => ({ right: 100 }));

    // сбрасываем состояние так чтобы метод прошел все условия внутри
    wrapper.instance().resetVisibleNodesData();
    wrapper.update();

    // вызываем метод
    wrapper.instance().checkVisibleNodes();
    wrapper.update();

    expect(wrapper.state()).toMatchObject({
      hasHiddenNodes: true,
      lastVisibleChildIndex: 4,
    });

    /*
     * Проверяем ситуацию когда все элементы модификаторов ВЛЕЗЛИ в высоту контейнера
     */
    Object.defineProperty(containerEl.children[5], 'offsetTop', {
      value: containerEl.offsetTop,
    });

    // сбрасываем состояние так чтобы метод прошел все условия внутри
    wrapper.instance().resetVisibleNodesData();
    wrapper.update();

    // вызываем метод
    wrapper.instance().checkVisibleNodes();
    wrapper.update();

    expect(wrapper.state()).toMatchObject({
      hasHiddenNodes: true,
      lastVisibleChildIndex: 4,
    });
  });

  it('checkVisibleNodes() set state hasHiddenNodes to false', () => {
    const wrapper = mount(
      <ModifiersGroup
        items={getTestItems()}
      />
    );

    const containerEl = wrapper.instance().containerRef.current;
    const showingButtonEl = wrapper.instance().showingButtonRef.current;

    Object.defineProperty(containerEl, 'clientHeight', {
      value: 200,
    });
    Object.defineProperty(containerEl, 'offsetTop', {
      value: 10,
    });

    Object.defineProperty(showingButtonEl, 'clientHeight', {
      value: 50,
    });
    Object.defineProperty(showingButtonEl, 'clientWidth', {
      value: 50,
    });

    [...containerEl.children].forEach((child, index) => {
      Object.defineProperty(child, 'clientHeight', {
        value: 50,
      });
      Object.defineProperty(child, 'clientWidth', {
        value: 50,
      });

      if (index >= 5) {
        Object.defineProperty(child, 'offsetTop', {
          value: 1,
          configurable: true,
        });
      }
    });

    wrapper.instance().resetVisibleNodesData();
    wrapper.update();

    wrapper.instance().checkVisibleNodes();
    wrapper.update();

    expect(wrapper.state()).toMatchObject({
      hasHiddenNodes: false,
      lastVisibleChildIndex: null,
    });
  });

  it('checkVisibleNodes(), expect isShowingButtonFit return 1', () => {
    const wrapper = mount(
      <ModifiersGroup
        items={getTestItems()}
      />
    );
    const containerEl = wrapper.instance().containerRef.current;
    const showingButtonEl = wrapper.instance().showingButtonRef.current;

    // подменяем значения высоты и отступа сверху у контейнера
    Object.defineProperty(containerEl, 'clientHeight', {
      value: 200,
    });
    Object.defineProperty(containerEl, 'offsetTop', {
      value: 10,
    });

    // подменяем значения высоты и ширины у кнопки показа всех модификаторов
    Object.defineProperty(showingButtonEl, 'clientHeight', {
      value: 50,
    });
    Object.defineProperty(showingButtonEl, 'clientWidth', {
      value: 50,
    });

    // подменяем значения у дочерних элементов (модификаторов) контейнера
    [...containerEl.children].forEach((child, index) => {
      Object.defineProperty(child, 'clientHeight', {
        value: 50,
      });
      Object.defineProperty(child, 'clientWidth', {
        value: 50,
      });

      jest.spyOn(containerEl, 'getBoundingClientRect').mockImplementation(() => ({ right: 200 }));
      jest.spyOn(containerEl.children[4], 'getBoundingClientRect').mockImplementation(() => ({ right: 100 }));

      /*
       * после 5 элемента подменяем у всех элементов offsetTop так,
       * чтобы компоненту ModifiersGroup казалось что он не влез и перенесся
       */
      if (index >= 5) {
        Object.defineProperty(child, 'offsetTop', {
          value: 220,
          configurable: true,
        });
      }
    });

    /*
     * Проверяем ситуацию когда один элемент не влез в высоту контейнера
     * и кнопке НЕ ХВАТИЛО место после последнего влезшего
     */

    // сбрасываем состояние так чтобы метод прошел все условия внутри
    wrapper.instance().resetVisibleNodesData();
    wrapper.update();

    // вызываем метод
    wrapper.instance().checkVisibleNodes();
    wrapper.update();

    expect(wrapper.state()).toMatchObject({
      hasHiddenNodes: true,
      lastVisibleChildIndex: 5,
    });

    // сбрасываем состояние так чтобы метод прошел все условия внутри
    wrapper.instance().resetVisibleNodesData();
    wrapper.update();

    // вызываем метод
    wrapper.instance().checkVisibleNodes();
    wrapper.update();

    expect(wrapper.state()).toMatchObject({
      hasHiddenNodes: true,
      lastVisibleChildIndex: 5,
    });

    /*
     * Проверяем ситуацию когда все элементы модификаторов ВЛЕЗЛИ в высоту контейнера
     */
    Object.defineProperty(containerEl.children[5], 'offsetTop', {
      value: containerEl.offsetTop,
    });

    // сбрасываем состояние так чтобы метод прошел все условия внутри
    wrapper.instance().resetVisibleNodesData();
    wrapper.update();

    // вызываем метод
    wrapper.instance().checkVisibleNodes();
    wrapper.update();

    expect(wrapper.state()).toMatchObject({
      hasHiddenNodes: true,
      lastVisibleChildIndex: 5,
    });
  });

  it('showing button click should update state properly', () => {
    const wrapper = mount(
      <ModifiersGroup
        items={getTestItems()}
      />
    );

    expect(wrapper.find(ModifierButton)).toHaveLength(6);
    expect(wrapper.find(ModifierButton).last().prop('className')).toContain('show-all-button');
    wrapper.find(ModifierButton).last().simulate('click');
    expect(wrapper.state()).toMatchObject({
      needShowAll: true,
    });
  });

  it('should handle "onSelectItem" prop', () => {
    const spy = jest.fn();
    const wrapper = mount(
      <ModifiersGroup
        onSelectItem={spy}
        items={getTestItems()}
      />
    );
    expect(spy).toHaveBeenCalledTimes(0);
    wrapper.find(ModifierButton).first().simulate('click');
    expect(spy).toHaveBeenCalledTimes(1);
    expect(spy.mock.calls[0][0]).toEqual({ content: 'Test 1' });
  });

  it('should handle "getItemType" prop', () => {
    const spy = jest.fn(() => 'test-type');
    const testItems = getTestItems();
    const wrapper = mount(
      <ModifiersGroup
        getItemType={spy}
        items={getTestItems()}
      />
    );
    testItems.forEach((item, index) => {
      expect(wrapper.find(ModifierButton).at(index).prop('type')).toBe('test-type');
    });
  });

  it('should handle "getItemCount" prop', () => {
    const spy = jest.fn(() => 'test-count');
    const testItems = getTestItems();
    const wrapper = mount(
      <ModifiersGroup
        getItemCount={spy}
        items={getTestItems()}
      />
    );
    testItems.forEach((item, index) => {
      expect(wrapper.find(ModifierButton).at(index).prop('count')).toBe('test-count');
    });
  });

  it('should handle "getItemColor" prop', () => {
    const spy = jest.fn(() => 'test-color');
    const testItems = getTestItems();
    const wrapper = mount(
      <ModifiersGroup
        getItemColor={spy}
        items={getTestItems()}
      />
    );
    testItems.forEach((item, index) => {
      expect(wrapper.find(ModifierButton).at(index).prop('color')).toBe('test-color');
    });
  });

  it('should handle "isSelectedItem" prop', () => {
    const spy = jest.fn(() => 'test-selected');
    const testItems = getTestItems();
    const wrapper = mount(
      <ModifiersGroup
        isSelectedItem={spy}
        items={getTestItems()}
      />
    );
    testItems.forEach((item, index) => {
      expect(wrapper.find(ModifierButton).at(index).prop('selected')).toBe('test-selected');
    });
  });

  it('should handle "isMarkdown" prop', () => {
    const spy = jest.fn(() => 'testMarkdown');
    const testItems = getTestItems();
    const wrapper = mount(
      <ModifiersGroup
        items={getTestItems()}
        isMarkdown={spy}
      />
    );
    testItems.forEach((item, index) => {
      expect(wrapper.find(ModifierButton).at(index).prop('isMarkdown')).toBe('testMarkdown');
    });
  });

  it('should match snapshot', () => {
    const testItems = getTestItems().map((item, index) => ({
      ...item,
      chosen: index === 2,
      bgColor: index === 2 ? '#f00' : undefined,
      text: index > 2 ? 'test text' : undefined,
      image: index < 2 ? 'www.test-image.com' : undefined,
      amount: index,
    }));
    const wrapper = shallow(
      <ModifiersGroup
        items={testItems}
        getItemType={item => item.image ? 'image' : 'text'}
        getItemContent={item => item.text || item.image}
        getItemImage={prop('image')}
        getItemCount={prop('amount')}
        getItemColor={prop('bgColor')}
        isSelectedItem={prop('chosen')}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should show all modifiers, if pass prop needShowAll', () => {
    const component = mount(
      <ModifiersGroup
        items={getTestItems()}
        needShowAll
      />
    );
    component.setState({ lastVisibleChildIndex: 3 });
    expect(component.find(ModifierButton).filter('.show-all-button')).toHaveLength(0);
    expect(component.find(ModifierButton).children().filter('.not-display')).toHaveLength(0);
  });

  it('should call onClickShowAll if it is a function', function () {
    const spy = jest.fn();
    const component = shallow(
      <ModifiersGroup
        items={getTestItems()}
        onClickShowAll={spy}
      />
    );
    component.setState({ lastVisibleChildIndex: 3 });
    expect(spy).not.toBeCalled();
    component.find(ModifierButton).last().simulate('click');
    expect(spy).toBeCalledTimes(1);
  });
});

describe('createHiddenNodeIndexFinder()', () => {
  it('should return false', () => {
    const find = createHiddenNodeIndexFinder({
      offsetTop: 0,
      clientHeight: 100,
    });
    const node = {
      offsetTop: 0,
      classList: { contains: () => false },
    };
    const nodeInvisible = {
      ...node,
      classList: { contains: () => true },
    };

    expect(find(node)).toBe(false);
    expect(find(nodeInvisible)).toBe(false);
  });

  it('should return true', () => {
    const find = createHiddenNodeIndexFinder({
      offsetTop: 0,
      clientHeight: 100,
    });
    const node = {
      offsetTop: 120,
      classList: { contains: () => false },
    };
    const nodeInvisible = {
      ...node,
      classList: { contains: () => true },
    };

    expect(find(node)).toBe(true);
    expect(find(nodeInvisible)).toBe(true);
  });
});
