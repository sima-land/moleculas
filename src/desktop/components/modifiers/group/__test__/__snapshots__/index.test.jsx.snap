// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ModifiersGroup /> should match snapshot 1`] = `
<div
  className="modifiers-group height-limited"
>
  <ForwardRef(ModifierButton)
    className="guttered"
    content="www.test-image.com"
    count={0}
    image="www.test-image.com"
    key="0"
    onClick={[Function]}
    selected={false}
    type="image"
  />
  <ForwardRef(ModifierButton)
    className="guttered"
    content="www.test-image.com"
    count={1}
    image="www.test-image.com"
    key="1"
    onClick={[Function]}
    selected={false}
    type="image"
  />
  <ForwardRef(ModifierButton)
    className="guttered"
    color="#f00"
    count={2}
    key="2"
    onClick={[Function]}
    selected={true}
    type="text"
  />
  <ForwardRef(ModifierButton)
    className="guttered"
    content="test text"
    count={3}
    key="3"
    onClick={[Function]}
    selected={false}
    type="text"
  />
  <ForwardRef(ModifierButton)
    className="guttered"
    content="test text"
    count={4}
    key="4"
    onClick={[Function]}
    selected={false}
    type="text"
  />
  <ForwardRef(ModifierButton)
    className="show-all-button guttered invisible"
    content="+5"
    onClick={[Function]}
    squared={true}
  />
</div>
`;
